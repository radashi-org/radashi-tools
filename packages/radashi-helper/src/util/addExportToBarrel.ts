/**
 * This function assumes the `content` represents a barrel file, which
 * simply re-exports a list of modules. It also assumes you want to
 * export everything from the given `specifier`. Lastly, it assumes
 * the barrel is alphabetically sorted and all lines begin with
 * `export * from` and use single quotes for the specifiers.
 *
 * In other words, this function is brittle and purpose-built.
 */
export function addExportToBarrel(content: string, specifier: string) {
  const addedLine = `export * from '${specifier}'`
  const existingLines = content.split('\n')

  let insertIndex = Math.max(existingLines.length - 1, 0)

  for (let i = 0; i < existingLines.length; i++) {
    if (existingLines[i] && existingLines[i] > addedLine) {
      if (i === 0) {
        insertIndex = 0
        break
      }
      for (let j = i - 1; j >= 0; j--) {
        if (existingLines[j].length) {
          insertIndex = j + 1
          break
        }
      }
      break
    }
  }

  existingLines.splice(insertIndex, 0, addedLine)

  return existingLines.join('\n')
}
