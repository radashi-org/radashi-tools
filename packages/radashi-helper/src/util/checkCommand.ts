import { exec } from 'node:child_process'

export async function checkCommand(cmd: string) {
  return new Promise<boolean>(resolve => {
    exec(`command -v ${cmd}`, error => {
      resolve(!error)
    })
  })
}
