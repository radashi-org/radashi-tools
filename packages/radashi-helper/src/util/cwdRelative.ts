import { white } from 'kleur/colors'
import { relative } from 'node:path'

export function cwdRelative(path: string) {
  let result = relative(process.cwd(), path)
  if (!result.startsWith('..')) {
    result = `./${result}`
  }
  return white(result)
}
