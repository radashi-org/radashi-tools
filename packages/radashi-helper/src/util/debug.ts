import createDebug from 'debug'

export const debug = createDebug('radashi')
