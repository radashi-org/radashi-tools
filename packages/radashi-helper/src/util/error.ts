import { isError } from 'radashi'
import { log } from './log'

export class RadashiError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'RadashiError'
  }
}

export class EarlyExitError extends RadashiError {
  constructor(message = 'Cannot continue. Exiting...') {
    super(message)
    this.name = 'EarlyExitError'
  }
}

export function forwardStderrAndRethrow(error: any): never {
  if (isError(error) && 'stderr' in error) {
    if ('stdout' in error) {
      log(error.stdout as string)
    }
    log.error(error.stderr as string)
  }
  throw error
}
