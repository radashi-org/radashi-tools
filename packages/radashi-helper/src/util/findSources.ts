import glob from 'fast-glob'
import type { Env } from '../env'

const fileTypes = ['src', 'overrides', 'rewired'] as const

type FileType = (typeof fileTypes)[number]

export async function findSources<Only extends FileType = FileType>(
  env: Env,
  only?: Only[],
): Promise<{
  src: 'src' extends Only ? string[] : undefined
  overrides: 'overrides' extends Only ? string[] : undefined
  rewired: 'rewired' extends Only ? string[] : undefined
}>

export async function findSources(
  env: Env,
  only?: FileType[],
): Promise<{
  src: string[] | undefined
  overrides: string[] | undefined
  rewired: string[] | undefined
}> {
  const [src, overrides, rewired] = await Promise.all(
    [
      {
        id: 'src' as const,
        glob: ['src/**/*.ts', '!src/*.ts'],
      },
      {
        id: 'overrides' as const,
        glob: ['overrides/src/**/*.ts', '!overrides/src/*.ts'],
      },
      {
        id: 'rewired' as const,
        glob: ['overrides/rewired/**/*.ts'],
      },
    ].map(file =>
      only?.includes(file.id) !== false
        ? glob(file.glob, {
            cwd: env.root,
            absolute: true,
          })
        : undefined,
    ),
  )

  return {
    src,
    overrides,
    rewired,
  }
}
