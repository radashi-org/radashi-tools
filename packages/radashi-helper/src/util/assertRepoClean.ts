import { RadashiError } from './error'
import { isRepoClean } from './isRepoClean'

export async function assertRepoClean(cwd: string) {
  if (!(await isRepoClean(cwd))) {
    throw new RadashiError(
      'Your repository has uncommitted changes.' +
        ' Please commit or stash them before overriding.',
    )
  }
}
