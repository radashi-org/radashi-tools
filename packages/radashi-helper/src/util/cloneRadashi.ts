import { execa, type Options as ExecaOptions } from 'execa'
import type { PackageJson } from 'type-fest'
import { forwardStderrAndRethrow, RadashiError } from './error'
import { prepareGitHubDefaultRepo } from './github'

export async function getInstalledRadashiRef(pkg: PackageJson) {
  const specifier = pkg.dependencies?.radashi
  if (!specifier) {
    throw new RadashiError('No radashi dependency found in package.json')
  }

  let ref = specifier.split('@').pop()!

  if (ref === 'beta') {
    ref = 'main'
  } else if (ref !== 'next') {
    ref = await execa('npm', ['view', 'radashi@' + ref, '--json']).then(
      ({ stdout }) => 'v' + JSON.parse(stdout).version,
    )
  }

  return ref
}

export function isPreReleaseRadashiTag(ref: string) {
  return /-(beta|alpha)/.test(ref)
}

export function getRadashiCloneURL(ref: string) {
  let cloneUrl = 'https://github.com/radashi-org/radashi'
  if (isPreReleaseRadashiTag(ref)) {
    // The "radashi-beta" repo is where versioned pre-release tags are
    // stored, which helps to avoid spamming the main repo.
    cloneUrl += '-beta'
  }
  return cloneUrl
}

export async function cloneRadashi(
  ref: string,
  dir: string,
  opts: ExecaOptions = {},
) {
  const cloneUrl = getRadashiCloneURL(ref)
  await execa(
    'git',
    ['clone', cloneUrl, '--branch', ref, '--single-branch', dir],
    opts,
  ).catch(error => {
    if (opts.stdio == null) {
      forwardStderrAndRethrow(error)
    }
    throw error
  })
  await prepareGitHubDefaultRepo(dir)
}
