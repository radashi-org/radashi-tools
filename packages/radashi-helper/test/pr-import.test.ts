import {
  expectGitDiffToMatchSnapshot,
  expectLogsToMatchSnapshot,
} from './util/expect'
import { getGitCommitSHAs } from './util/git'
import { debug } from './util/mocks'
import { run } from './util/run'
import { replaceWordsInString } from './util/string'

test('import a pr', async () => {
  debug()

  await run('pr import 208')

  // Commit hashes are non-deterministic, so we replace them with a
  // placeholder.
  const commits = await getGitCommitSHAs({ short: true })
  expectLogsToMatchSnapshot(logs =>
    replaceWordsInString(logs, commits, '*******'),
  )

  await expectGitDiffToMatchSnapshot('last-commit')
})
