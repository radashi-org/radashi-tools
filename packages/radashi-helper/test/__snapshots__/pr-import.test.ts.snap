// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`import a pr > git diff (last-commit) 1`] = `
"diff --git a/benchmarks/number/spiralSum.bench.ts b/benchmarks/number/spiralSum.bench.ts
new file mode 100644
index 0000000..2fe6664
--- /dev/null
+++ b/benchmarks/number/spiralSum.bench.ts
@@ -0,0 +1,7 @@
+import * as _ from 'radashi'
+
+describe('spiralSum', () => {
+  bench('calculate spiral sum for n=5', () => {
+    _.spiralSum(5)
+  })
+})
diff --git a/docs/number/spiralSum.mdx b/docs/number/spiralSum.mdx
new file mode 100644
index 0000000..1c2d6f9
--- /dev/null
+++ b/docs/number/spiralSum.mdx
@@ -0,0 +1,39 @@
+---
+title: spiralSum
+description: Calculates the sum of numbers in a spiral pattern
+---
+
+### Usage
+
+The \`spiralSum\` function calculates the sum of numbers in a unique spiral pattern up to a specified number of terms.
+
+\`\`\`ts
+import { spiralSum } from 'radashi'
+
+spiralSum(5) // => 25
+\`\`\`
+
+### How It Works
+
+The spiral pattern is formed by summing consecutive odd numbers:
+
+1. First term: 1
+2. Second term: 1 + 3 = 4
+3. Third term: 1 + 3 + 5 = 9
+4. Fourth term: 1 + 3 + 5 + 7 = 16
+5. Fifth term: 1 + 3 + 5 + 7 + 9 = 25
+
+And so on...
+
+### Edge Cases
+
+- If \`n\` is 0 or negative, the function returns 0.
+
+\`\`\`ts
+spiralSum(0) // => 0
+spiralSum(-1) // => 0
+\`\`\`
+
+## Notes
+
+This function is primarily used for testing and demonstration purposes. It provides a simple way to generate an interesting numerical sequence based on a spiral pattern of odd numbers.
diff --git a/mod.ts b/mod.ts
index 1807e49..bfa483f 100644
--- a/mod.ts
+++ b/mod.ts
@@ -4,3 +4,4 @@ export { AggregateError, Any, type Assign, type BoxedPrimitive, type BuiltInType
 
 // Our custom functions.
 export { add } from './src/number/add.ts'
+export { spiralSum } from './src/number/spiralSum.ts'
diff --git a/src/number/spiralSum.ts b/src/number/spiralSum.ts
new file mode 100644
index 0000000..578d552
--- /dev/null
+++ b/src/number/spiralSum.ts
@@ -0,0 +1,32 @@
+/**
+ * Calculates the sum of numbers in a spiral pattern up to n terms.
+ *
+ * The spiral pattern goes: 1, 1+3, 1+3+5, 1+3+5+7, 1+3+5+7+9, ...
+ * Where each term adds the next odd number to the previous sum.
+ *
+ * @example
+ * \`\`\`typescript
+ * spiralSum(1) // => 1
+ * spiralSum(2) // => 4
+ * spiralSum(3) // => 9
+ * spiralSum(4) // => 16
+ * spiralSum(5) // => 25
+ * \`\`\`
+ */
+export function spiralSum(n: number): number {
+  if (n <= 0) {
+    return 0
+  }
+
+  let total = 1
+  let current = 1
+  let increment = 1
+
+  for (let i = 1; i < n; i++) {
+    increment += 2
+    current += increment
+    total += current
+  }
+
+  return total
+}
diff --git a/tests/number/spiralSum.test.ts b/tests/number/spiralSum.test.ts
new file mode 100644
index 0000000..f4a7154
--- /dev/null
+++ b/tests/number/spiralSum.test.ts
@@ -0,0 +1,17 @@
+import * as _ from 'radashi'
+
+describe('spiralSum', () => {
+  it('calculates the correct spiral sum for various inputs', () => {
+    expect(_.spiralSum(0)).toBe(0)
+    expect(_.spiralSum(1)).toBe(1)
+    expect(_.spiralSum(2)).toBe(4)
+    expect(_.spiralSum(3)).toBe(9)
+    expect(_.spiralSum(4)).toBe(16)
+    expect(_.spiralSum(5)).toBe(25)
+  })
+
+  it('returns 0 for negative inputs', () => {
+    expect(_.spiralSum(-1)).toBe(0)
+    expect(_.spiralSum(-100)).toBe(0)
+  })
+})"
`;

exports[`import a pr > logs 1`] = `
"[info] Checking out PR...
[info] 
[info] [test *******] feat: add \`spiralSum\` function
[info]  Author: Radashi Bot <175859458+radashi-bot@users.noreply.github.com>
[info]  5 files changed, 96 insertions(+)
[info]  create mode 100644 benchmarks/number/spiralSum.bench.ts
[info]  create mode 100644 docs/number/spiralSum.mdx
[info]  create mode 100644 src/number/spiralSum.ts
[info]  create mode 100644 tests/number/spiralSum.test.ts"
`;
